### ISSUE TITLE
Test selectors should be applied to elements having ...attributes

### ISSUE THREAD

**Issue by jelhan** [simplabs/ember-test-selectors#383]
TL;DR: Tagless components should be supported in curly bracket component invocation syntax by applying test selectors to the element that has `...attributes`.

If I have a tagless component `<button ...attributes>{{yield}}</button>` and invoke it like `{{#my-button data-test-foo=true}}...{{/my-button}}`, the rendered HTML should be `<button data-test-foo>...</button>`.

An assertion should only be thrown if a tagless component does not apply HTML attributes to any element (does not use `...attributes`).

---

Currently Ember Test Selectors does not support tagless components in curly brackets / classic component invocation syntax. It throws an assertion if tried so.

This limitations makes incremental adoption of outerHTML pattern difficult. All component invocations that uses test selectors must be converted to angle bracket syntax before the component is refactored to a tagless component internally. This is especially an issue if the component is provided by an addon.

Consider a simple button component which might have been like this for a long time:

```js
export default Component.extend({
  tagName: 'button'
});
```

```hbs
{{yield}}
```

The maintainer now decides to adopt outerHTML pattern as that's quite famous with Ember Octance and fits better with element modifiers etc. He refactors the component to something like:

```js
export default Component.extend({
  tagName: ''
});
```

```hbs
<button ...attributes>{{yield}}</button>
```

As this isn't breaking the API scope of that component it's considered an internal change and released with a minor version. But as Ember Test Selector relies on the internal structure of that component the tests of an application will fail if using test selectors on that component and haven't refactored to angle bracket invocation syntax yet.

As Ember Test Selectors is a quite famous addon and used by a lot of applications this would be very bad for the hole ecosystem and the developer experience as soon as more and more addons refactor for tagless components.

I don't think we are facing a technical limitation here. It's not that easy but it should be doable.

My first idea was applying an element modifiers to all elements having `...attributes` using an AST transform at build time. That element modifiers gets the test selector properties of the component as positional arguments and is responsible for setting the data attributes on the element. I then remembered that element modifiers do not run in SSR / FastBoot...

Another approach might be adding a selector to all elements having `...attributes` using an AST transform. That selector could be used later in `didRender` hooks to set the test selectors on the element. This is a very basic implementation I came up with:

```js
export default Component.extend({
  tagName: '',

  // `attributes-target` data attribute should be applied to all elements
  // having `...attributes` at build time using an AST transform
  layout: hbs`<button data-attributes-target ...attributes></button>`,

  // using an element modifier would be more straight-forward and easier to
  // implement but isn't an option cause element modifiers are not executed
  // in server-side rendering / FastBoot.
  didRender() {
    this._super(...arguments);

    // https://stackoverflow.com/questions/51391654/is-there-a-way-to-get-all-the-attributes-passed-to-an-ember-component/51392135#51392135
    // eslint-disable-next-line ember/no-attrs-in-components
    let testSelectors = Object.keys(this.attrs).filter((key) => key.startsWith('data-test-'));

    // https://stackoverflow.com/questions/50112121/ember-tagless-component-innerhtml
    let range = Ember.ViewUtils.getViewBounds(this);

    // TODO: Simple-dom used by FastBoot does not support `querySelectorAll`. Must loop over
    //       all silbings of `firstNode` in range and there childrens to find all elements
    //       having the data attribute.
    let attributesTargets = range.parentElement.querySelectorAll('[data-attributes-target]');

    Array.from(attributesTargets).forEach((el) => {
      testSelectors.forEach((testSelector) => {
        el.setAttribute(testSelector, this[testSelector]);
      })
    });
  },

  supportsDataTestProperties: true,
})
```

Here are some tests you could use to play around with that approach. It's also available in this CodeSandbox: https://codesandbox.io/s/my-app-40w45?fontsize=14&initialpath=%2Ftests

```js
import { module, test } from "qunit";
import { setupRenderingTest } from "ember-qunit";
import { render } from "@ember/test-helpers";
import hbs from "htmlbars-inline-precompile";
import Component from "@ember/component";
/* global Ember */

module("Integration | Component | showcase", function(hooks) {
  setupRenderingTest(hooks);

  module("inner HTML", function(hooks) {
    hooks.beforeEach(function() {
      this.owner.register(
        "component:my-component",
        Component.extend({
          tagName: "button",
          layout: hbs``
        })
      );
    });

    test("classic component invocation syntax", async function(assert) {
      await render(hbs`{{my-component data-test-foo=true}}`);
      assert.dom("button").hasAttribute("data-test-foo");
    });

    test("angle bracket component invocation syntax", async function(assert) {
      await render(hbs`<MyComponent data-test-foo />`);
      assert.dom("button").hasAttribute("data-test-foo");
    });
  });

  module("outer HTML", function(hooks) {
    hooks.beforeEach(function() {
      this.owner.register(
        "component:my-component",
        Component.extend({
          tagName: "",
          layout: hbs`<button ...attributes></button>`
        })
      );
    });

    // this test is expected to be broken due to ember-test-selector assertion
    test("classic component invocation syntax", async function(assert) {
      await render(hbs`{{my-component data-test-foo=true}}`);
      assert.dom("button").hasAttribute("data-test-foo");
    });

    test("angle bracket component invocation syntax", async function(assert) {
      await render(hbs`<MyComponent data-test-foo />`);
      assert.dom("button").hasAttribute("data-test-foo");
    });
  });

  module("outer HTML fixed", function(hooks) {
    hooks.beforeEach(function() {
      this.owner.register(
        "component:my-component",
        Component.extend({
          tagName: "",

          // `attributes-target` data attribute should be applied to all elements
          // having `...attributes` at build time using an AST transform
          layout: hbs`<button data-attributes-target ...attributes></button>`,

          // using an element modifier would be more straight-forward and easier to
          // implement but isn't an option cause element modifiers are not executed
          // in server-side rendering / FastBoot.
          didRender() {
            this._super(...arguments);

            // https://stackoverflow.com/questions/51391654/is-there-a-way-to-get-all-the-attributes-passed-to-an-ember-component/51392135#51392135
            // eslint-disable-next-line ember/no-attrs-in-components
            let testSelectors = Object.keys(this.attrs).filter(key =>
              key.startsWith("data-test-")
            );

            // https://stackoverflow.com/questions/50112121/ember-tagless-component-innerhtml
            let range = Ember.ViewUtils.getViewBounds(this);

            // TODO: Simple-dom used by FastBoot does not support `querySelectorAll`. Must loop over
            //       all silbings of `firstNode` in range and there childrens to find all elements
            //       having the data attribute.
            let attributesTargets = range.parentElement.querySelectorAll(
              "[data-attributes-target]"
            );

            Array.from(attributesTargets).forEach(el => {
              testSelectors.forEach(testSelector => {
                el.setAttribute(testSelector, this[testSelector]);
              });
            });
          },

          supportsDataTestProperties: true
        })
      );
    });

    test("classic component invocation syntax", async function(assert) {
      await render(hbs`{{my-component data-test-foo=true}}`);
      assert.dom("button").hasAttribute("data-test-foo");
    });

    test("angle bracket component invocation syntax", async function(assert) {
      await render(hbs`<MyComponent data-test-foo />`);
      assert.dom("button").hasAttribute("data-test-foo");
    });
  });
});
```

I don't think that's the best solution possible but wanted to get the discussion started. I think we should fix that issue before it hurts to much.

**Comment by Turbo87** [comment]
@jelhan implementing this would significantly change the way that things are passed through in components and I'd like to avoid such surprises. My suggestion is to move to angle brackets first and then use tagless components. Doing it the other way around will cause issues like you're seeing right now, and not just for test-selectors but other things as well.

**Comment by jelhan** [comment]
> My suggestion is to move to angle brackets first and then use tagless components. Doing it the other way around will cause issues like you're seeing right now, and not just for test-selectors but other things as well.

That doesn't work out for components provided by 3rd-party addons. That would either mean that addons should not adopt tagless component pattern before support for curly bracket component invocation syntax has been deprecated and removed (unforeseeable future) or that applications should upgrade all usages of components provided by 3rd-party addons to angle bracket component invocation syntax before upgrading the addon. The latter one would likely cause friction of the ecosystem. It would also cause bad developer experience cause addon maintainers might not consider supporting test selectors as part of their public API and therefore break them in a minor release.

**Comment by marcoow** [comment]
So since curly invocation does not work with `...attributes` anyway (unless I'm missing something?), I don't see how this is an issue with ember-test-selectors. I understand the point about friction but if we somehow added special behavior for `data-test-` attributes so you could actually use them with curly invocation and `...attributes` and they would actually show up on the DOM element and thus behave differently than **all** other values you're passing to a component, that would actually cause **more** friction.

I suggest we close this.

**Comment by jelhan** [comment]
> I understand the point about friction but if we somehow added special behavior for `data-test-` attributes so you could actually use them with curly invocation and `...attributes` and they would actually show up on the DOM element and thus behave differently than **all** other values you're passing to a component, that would actually cause **more** friction.

This is taking the wrong perspective in my opinion. `ember-test-selectors` has always added special behavior to components. It was mapping properties which starts with `data-test-` to HTML attributes on the element defined by the component. For a long time there was only one common way to do such a mapping: `attributeBindings`. But that has changed with angle bracket component invocation syntax. Therefor `ember-test-selectors` should also support that new way - especially cause it will be the recommended one shortly.

It's not that `ember-test-selectors` would alter the behavior of the components more than before. Lets take an example for the recommended transition path of a component from classic to tagless form:

Before a simple button might have been implemented like this:

```js
export default Component.extend({
  tagName: 'button',
  attributeBindings: ['buttonType:type', 'disabled'],
});
```

Lets refactor that component without introducing a breaking change to a tagless one.

```hbs
<button
  buttonType={{@type}}
  disabled={{@disabled}}
  ...attributes
>
```

The maintainer decided to still support the old properties `@type` and `@disabled` to not break backward compatibility. He has taken care of all properties he is supporting and is aware of. A consumer might not even notice that there was a change cause `{{my-button buttonType="submit" disabled=true}}` would work before as after - unless he is using test selectors with that component.

In the old form ember-test-selectors was mapping a property to a HTML attribute of component's element. It was similar as if the maintainer would have added that attribute to the `attributeBindings` array. I'm arguing that it should do the same afterwards: Mapping the property to a HTML attribute of component's element as if the maintainer has added `data-test-foo={{@data-test-foo}}` to that element. That the component's element is not defined by `tagName` of that component but by applying `...attributes` to an element in the template is an implementation detail in my perspective.

**Comment by marcoow** [comment]
I'm going to close this. Adding the suggested code to ETS would mean quite some added complexity and magic and we don't think that's worth it as you can also switch to angle bracket invocation. Even if you cannot switch to angle bracket invocation, you could still do something like this in your app:

```js
import Ember from 'ember';
import OriginalButtonComponent from 'addon/button';

export default OriginalButtonComponent.extend({
  supportsDataTestProperties: true,

  didRender() {
    this._super(...arguments);

    let dataTestAttributes = Object.keys(this.attrs).filter((key) => key.startsWith('data-test-'));
    let range = Ember.ViewUtils.getViewBounds(this);
    let element = range.parentElement.querySelector('button');
    dataTestAttributes.forEach((attribute) => {
      element.setAttribute(attribute, this[attribute]);
    });
  }
});
```

While this is tedious and not great, I think that's good enough if it allows us to not add a substantial amount of magic to ETS (that we'd then also have to keep around for the long term).
